File: src/environment.ts
import fs from 'node:fs';
import os from 'node:os';
import path from 'node:path';

interface ErrnoException extends Error {
  code?: string;
}

export type EnvironmentDiagnosticCode =
  | 'model_cache_stat_failed'
  | 'model_cache_not_directory'
  | 'model_cache_creation_failed'
  | 'model_cache_unwritable'
  | 'model_cache_unavailable';

export interface EnvironmentDiagnostic {
  readonly level: 'warn' | 'error';
  readonly code: EnvironmentDiagnosticCode;
  readonly message: string;
  readonly context?: Record<string, unknown>;
}

export type ModelCacheSource =
  | 'FASTEMBED_CACHE_DIR'
  | 'INDEX_MCP_MODEL_CACHE_DIR'
  | 'default'
  | 'tmp';

interface DirectoryCandidate {
  readonly path: string;
  readonly source: ModelCacheSource;
}

const diagnostics: EnvironmentDiagnostic[] = [];
const recordedFailures = new Set<string>();

let modelCache: { directory: string; source: ModelCacheSource } | null = null;

function toAbsolutePath(candidate: string): string {
  return path.isAbsolute(candidate) ? candidate : path.resolve(candidate);
}

function recordDiagnostic(diagnostic: EnvironmentDiagnostic): void {
  const key = [diagnostic.code, diagnostic.context?.source, diagnostic.context?.path]
    .filter(Boolean)
    .join('::');

  if (diagnostic.level === 'error' && key) {
    if (recordedFailures.has(key)) {
      return;
    }
    recordedFailures.add(key);
  }

  diagnostics.push(diagnostic);
}

function toDiagnosticContext(candidate: DirectoryCandidate, error: unknown): Record<string, unknown> {
  const context: Record<string, unknown> = {
    source: candidate.source,
    path: candidate.path
  };

  if (error && typeof error === 'object') {
    const err = error as ErrnoException;
    if (typeof err.code === 'string') {
      context.code = err.code;
    }
    if (typeof err.message === 'string') {
      context.error = err.message;
    }
  } else if (error) {
    context.error = String(error);
  }

  return context;
}

function prepareDirectory(candidate: DirectoryCandidate): boolean {
  try {
    const stats = fs.statSync(candidate.path);
    if (!stats.isDirectory()) {
      recordDiagnostic({
        level: 'error',
        code: 'model_cache_not_directory',
        message: `[index-mcp] Model cache path ${candidate.path} exists but is not a directory.`,
        context: { source: candidate.source, path: candidate.path }
      });
      return false;
    }
  } catch (error) {
    const maybeErrno = error as ErrnoException;
    if (maybeErrno?.code && maybeErrno.code !== 'ENOENT') {
      recordDiagnostic({
        level: 'error',
        code: 'model_cache_stat_failed',
        message: `[index-mcp] Unable to inspect model cache directory ${candidate.path}.`,
        context: toDiagnosticContext(candidate, error)
      });
      return false;
    }
  }

  try {
    fs.mkdirSync(candidate.path, { recursive: true });
  } catch (error) {
    recordDiagnostic({
      level: 'error',
      code: 'model_cache_creation_failed',
      message: `[index-mcp] Unable to create model cache directory ${candidate.path}.`,
      context: toDiagnosticContext(candidate, error)
    });
    return false;
  }

  try {
    fs.accessSync(candidate.path, fs.constants.W_OK);
  } catch (error) {
    recordDiagnostic({
      level: 'error',
      code: 'model_cache_unwritable',
      message: `[index-mcp] Model cache directory ${candidate.path} is not writable.`,
      context: toDiagnosticContext(candidate, error)
    });
    return false;
  }

  return true;
}

function getDefaultCandidates(): DirectoryCandidate[] {
  const candidates: DirectoryCandidate[] = [];
  const envPath = process.env.INDEX_MCP_MODEL_CACHE_DIR?.trim();
  if (envPath) {
    candidates.push({
      source: 'INDEX_MCP_MODEL_CACHE_DIR',
      path: toAbsolutePath(envPath)
    });
  }

  const homeDir = os.homedir();
  if (homeDir && homeDir.trim()) {
    candidates.push({
      source: 'default',
      path: path.join(homeDir, '.index-mcp', 'models')
    });
  }

  candidates.push({
    source: 'tmp',
    path: path.join(os.tmpdir(), 'index-mcp', 'models')
  });

  return candidates;
}

function setResolvedModelCache(candidate: DirectoryCandidate): void {
  modelCache = {
    directory: candidate.path,
    source: candidate.source
  };
  process.env.FASTEMBED_CACHE_DIR = candidate.path;
  if (!process.env.INDEX_MCP_MODEL_CACHE_DIR) {
    process.env.INDEX_MCP_MODEL_CACHE_DIR = candidate.path;
  }
}

export function ensureModelCacheDirectory(): void {
  if (modelCache) {
    return;
  }

  const fastembedEnv = process.env.FASTEMBED_CACHE_DIR?.trim();
  if (fastembedEnv) {
    const candidate: DirectoryCandidate = {
      source: 'FASTEMBED_CACHE_DIR',
      path: toAbsolutePath(fastembedEnv)
    };

    if (prepareDirectory(candidate)) {
      setResolvedModelCache(candidate);
      return;
    }

    // Fall through to defaults if the explicit FASTEMBED cache directory is unusable.
  }

  const candidates = getDefaultCandidates();

  for (const candidate of candidates) {
    if (prepareDirectory(candidate)) {
      setResolvedModelCache(candidate);
      return;
    }
  }

  recordDiagnostic({
    level: 'error',
    code: 'model_cache_unavailable',
    message:
      '[index-mcp] Unable to configure a writable embedding model cache directory. The embedding pipeline may create per-process caches in the working directory.',
    context: { candidates: candidates.map((candidate) => candidate.path) }
  });
}

export function getModelCacheConfiguration(): { directory: string | null; source: ModelCacheSource | null } {
  ensureModelCacheDirectory();
  return modelCache
    ? { directory: modelCache.directory, source: modelCache.source }
    : { directory: null, source: null };
}

export function getEnvironmentDiagnostics(): EnvironmentDiagnostic[] {
  return diagnostics.map((diagnostic) => ({ ...diagnostic }));
}

export function getBudgetTokens(): number {
  const budgetEnv = process.env.INDEX_MCP_BUDGET_TOKENS?.trim();
  if (budgetEnv) {
    const parsed = parseInt(budgetEnv, 10);
    if (!isNaN(parsed) && parsed > 0) {
      return parsed;
    }
  }
  return 3000; // Default budget
}

File: crates/index-mcp-server/src/environment.rs
use std::collections::HashSet;
use std::fs::{self, OpenOptions};
use std::io;
use std::path::{Path, PathBuf};

use dirs::home_dir;
use rmcp::schemars::JsonSchema;
use serde::Serialize;
use serde_json::{json, Map, Value};

/// Level for environment diagnostics.
#[derive(Debug, Clone, Serialize, JsonSchema, PartialEq, Eq, Hash)]
#[serde(rename_all = "lowercase")]
pub enum DiagnosticLevel {
    Warn,
    Error,
}

/// Source for the resolved model cache directory.
#[derive(Debug, Clone, Serialize, JsonSchema, PartialEq, Eq, Hash)]
pub enum ModelCacheSource {
    #[serde(rename = "FASTEMBED_CACHE_DIR")]
    FastembedCacheDir,
    #[serde(rename = "INDEX_MCP_MODEL_CACHE_DIR")]
    IndexMcpModelCacheDir,
    #[serde(rename = "default")]
    Default,
    #[serde(rename = "tmp")]
    Tmp,
}

impl ModelCacheSource {
    fn as_str(&self) -> &'static str {
        match self {
            ModelCacheSource::FastembedCacheDir => "FASTEMBED_CACHE_DIR",
            ModelCacheSource::IndexMcpModelCacheDir => "INDEX_MCP_MODEL_CACHE_DIR",
            ModelCacheSource::Default => "default",
            ModelCacheSource::Tmp => "tmp",
        }
    }
}

/// Diagnostic describing configuration or filesystem issues discovered while
/// preparing the embedding model cache directory.
#[derive(Debug, Clone, Serialize, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct EnvironmentDiagnostic {
    pub level: DiagnosticLevel,
    pub code: String,
    pub message: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub context: Option<Map<String, Value>>,
}

/// Snapshot of the runtime environment relevant to embedding model caching.
#[derive(Debug, Clone, Serialize, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct ModelCacheInfo {
    pub directory: Option<String>,
    pub source: Option<ModelCacheSource>,
    pub diagnostics: Vec<EnvironmentDiagnostic>,
}

struct DirectoryCandidate {
    path: PathBuf,
    source: ModelCacheSource,
}

/// Resolve and prepare the directory used for caching embedding models. The
/// behaviour mirrors the legacy Node server so cached models remain
/// interchangeable between runtimes.
pub fn collect_model_cache_info() -> ModelCacheInfo {
    let mut diagnostics = Vec::new();
    let mut recorded_failures = HashSet::new();

    let mut resolved_candidate: Option<DirectoryCandidate> = None;

    if let Some(path) = std::env::var("FASTEMBED_CACHE_DIR")
        .ok()
        .and_then(normalize_non_empty)
    {
        let candidate = DirectoryCandidate {
            path: to_absolute_path(&path),
            source: ModelCacheSource::FastembedCacheDir,
        };
        if prepare_directory(&candidate, &mut diagnostics, &mut recorded_failures) {
            resolved_candidate = Some(candidate);
        }
    }

    if resolved_candidate.is_none() {
        for candidate in default_candidates() {
            if prepare_directory(&candidate, &mut diagnostics, &mut recorded_failures) {
                resolved_candidate = Some(candidate);
                break;
            }
        }
    }

    if let Some(candidate) = resolved_candidate {
        let directory = candidate.path.display().to_string();
        std::env::set_var("FASTEMBED_CACHE_DIR", &directory);
        if std::env::var("INDEX_MCP_MODEL_CACHE_DIR")
            .ok()
            .and_then(normalize_non_empty)
            .is_none()
        {
            std::env::set_var("INDEX_MCP_MODEL_CACHE_DIR", &directory);
        }

        ModelCacheInfo {
            directory: Some(directory),
            source: Some(candidate.source),
            diagnostics,
        }
    } else {
        diagnostics.push(EnvironmentDiagnostic {
            level: DiagnosticLevel::Error,
            code: "model_cache_unavailable".to_string(),
            message: "[index-mcp] Unable to configure a writable embedding model cache directory."
                .to_string(),
            context: None,
        });

        ModelCacheInfo {
            directory: None,
            source: None,
            diagnostics,
        }
    }
}

fn normalize_non_empty(value: String) -> Option<String> {
    let trimmed = value.trim().to_string();
    if trimmed.is_empty() {
        None
    } else {
        Some(trimmed)
    }
}

fn default_candidates() -> Vec<DirectoryCandidate> {
    let mut candidates = Vec::new();

    if let Some(path) = std::env::var("INDEX_MCP_MODEL_CACHE_DIR")
        .ok()
        .and_then(normalize_non_empty)
    {
        candidates.push(DirectoryCandidate {
            path: to_absolute_path(&path),
            source: ModelCacheSource::IndexMcpModelCacheDir,
        });
    }

    if let Some(home) = home_dir() {
        let candidate = home.join(".index-mcp").join("models");
        candidates.push(DirectoryCandidate {
            path: candidate,
            source: ModelCacheSource::Default,
        });
    }

    let tmp = std::env::temp_dir().join("index-mcp").join("models");
    candidates.push(DirectoryCandidate {
        path: tmp,
        source: ModelCacheSource::Tmp,
    });

    candidates
}

fn to_absolute_path(candidate: &str) -> PathBuf {
    let path = PathBuf::from(candidate);
    if path.is_absolute() {
        path
    } else if let Ok(cwd) = std::env::current_dir() {
        cwd.join(path)
    } else {
        path
    }
}

fn prepare_directory(
    candidate: &DirectoryCandidate,
    diagnostics: &mut Vec<EnvironmentDiagnostic>,
    recorded_failures: &mut HashSet<String>,
) -> bool {
    if ensure_directory(candidate, diagnostics, recorded_failures).is_err() {
        return false;
    }

    if let Err(err) = check_writable(&candidate.path) {
        record_diagnostic(
            diagnostics,
            recorded_failures,
            EnvironmentDiagnostic {
                level: DiagnosticLevel::Error,
                code: "model_cache_unwritable".to_string(),
                message: format!(
                    "[index-mcp] Model cache directory {} is not writable.",
                    candidate.path.display()
                ),
                context: Some(diagnostic_context(candidate, Some(&err))),
            },
        );
        return false;
    }

    true
}

fn ensure_directory(
    candidate: &DirectoryCandidate,
    diagnostics: &mut Vec<EnvironmentDiagnostic>,
    recorded_failures: &mut HashSet<String>,
) -> Result<(), io::Error> {
    match fs::metadata(&candidate.path) {
        Ok(metadata) => {
            if !metadata.is_dir() {
                record_diagnostic(
                    diagnostics,
                    recorded_failures,
                    EnvironmentDiagnostic {
                        level: DiagnosticLevel::Error,
                        code: "model_cache_not_directory".to_string(),
                        message: format!(
                            "[index-mcp] Model cache path {} exists but is not a directory.",
                            candidate.path.display()
                        ),
                        context: Some(diagnostic_context(candidate, None)),
                    },
                );
                return Err(io::Error::new(io::ErrorKind::Other, "not a directory"));
            }
        }
        Err(err) => {
            if err.kind() != io::ErrorKind::NotFound {
                record_diagnostic(
                    diagnostics,
                    recorded_failures,
                    EnvironmentDiagnostic {
                        level: DiagnosticLevel::Error,
                        code: "model_cache_stat_failed".to_string(),
                        message: format!(
                            "[index-mcp] Unable to inspect model cache directory {}.",
                            candidate.path.display()
                        ),
                        context: Some(diagnostic_context(candidate, Some(&err))),
                    },
                );
                return Err(err);
            }
        }
    }

    if let Err(err) = fs::create_dir_all(&candidate.path) {
        record_diagnostic(
            diagnostics,
            recorded_failures,
            EnvironmentDiagnostic {
                level: DiagnosticLevel::Error,
                code: "model_cache_creation_failed".to_string(),
                message: format!(
                    "[index-mcp] Unable to create model cache directory {}.",
                    candidate.path.display()
                ),
                context: Some(diagnostic_context(candidate, Some(&err))),
            },
        );
        return Err(err);
    }

    Ok(())
}

fn check_writable(path: &Path) -> Result<(), io::Error> {
    let probe_path = path.join(".mcp-write-test");
    match OpenOptions::new()
        .create(true)
        .write(true)
        .truncate(true)
        .open(&probe_path)
    {
        Ok(_) => {
            let _ = fs::remove_file(probe_path);
            Ok(())
        }
        Err(err) => Err(err),
    }
}

fn diagnostic_context(
    candidate: &DirectoryCandidate,
    error: Option<&io::Error>,
) -> Map<String, Value> {
    let mut context = Map::new();
    context.insert(
        "source".to_string(),
        Value::String(candidate.source.as_str().to_string()),
    );
    context.insert(
        "path".to_string(),
        Value::String(candidate.path.display().to_string()),
    );

    if let Some(err) = error {
        context.insert("error".to_string(), Value::String(err.to_string()));
        context.insert(
            "code".to_string(),
            Value::String(format!("{:?}", err.kind())),
        );
    }

    context
}

fn record_diagnostic(
    diagnostics: &mut Vec<EnvironmentDiagnostic>,
    recorded_failures: &mut HashSet<String>,
    diagnostic: EnvironmentDiagnostic,
) {
    if diagnostic.level == DiagnosticLevel::Error {
        let key = json!({
            "code": diagnostic.code,
            "source": diagnostic
                .context
                .as_ref()
                .and_then(|ctx| ctx.get("source"))
                .cloned(),
            "path": diagnostic
                .context
                .as_ref()
                .and_then(|ctx| ctx.get("path"))
                .cloned(),
        })
        .to_string();

        if !recorded_failures.insert(key) {
            return;
        }
    }

    diagnostics.push(diagnostic);
}
